### ISSUE TITLE
Questions about training and loss functions

### ISSUE THREAD

**Issue by ivanmag22** [MouseLand/cellpose#1120]
## In a nutshell
I would like to better understand how the loss function for training works, specifically:
- how *cellprob* and *gradient flows contributions* influence the loss
- what **weighted parameter λ** represents
  - which values it can cover
  - what happens if we increase or decrease its value

During training, I found that using **λ = 1** instead of the default λ = 5 improved performance in my case. I would like to understand why.

## Description
I am using Cellpose to detect some nuclei in my confocal images and to train it on my samples.
As regard training, I read in your paper (Cellpose v1) that there is a **weighted parameter λ** in the loss function, in particular inside gradient flows contribution; by default this weighted parameter is set to 5.
I am not using your train method or your GUI and human-in-the-loop approach because I do not have annotated data. Instead, I adopted an unsupervised technique where the model learns from masks and tries to generate similar ones.
For this kind of training I used your loss function, passing properly the values to the method. However, training sometimes fails—the model either does not detect objects or performs poorly, even when the loss decreases to 0. Specifically, gradient flows cannot overcome *flow_threshold = 0.4*. So, one time I had the intuition to set λ to 1 and it worked, since in my approach I am dealing with outputs and I thought that that two outputs should be close to each other.

### Code
in cellpose/cellpose/train.py
```
def _loss_fn_seg(lbl, y, device):
    """
    Calculates the loss function between true labels lbl and prediction y.

    Args:
        lbl (numpy.ndarray): True labels (cellprob, flowsY, flowsX).
        y (torch.Tensor): Predicted values (flowsY, flowsX, cellprob).
        device (torch.device): Device on which the tensors are located.

    Returns:
        torch.Tensor: Loss value.

    """
    criterion = nn.MSELoss(reduction="mean")
    criterion2 = nn.BCEWithLogitsLoss(reduction="mean")
    veci = 5.0 * torch.from_numpy(lbl[:, 1:]).to(device)
    loss = criterion(y[:, :2], veci)
    loss /= 2.0
    loss2 = criterion2(y[:, -1], torch.from_numpy(lbl[:, 0] > 0.5).to(device).float())
    loss = loss + loss2
    return loss
```
where **weighted parameter λ** is present in ```veci``` as a factor. The two losses are summed with a weight of 1/2 on gradient flows contribution.

## Questions

1) why is this parameter **inside** the loss function and not **between** the two contributions, like l1 + w * l2 (where l1 and l2 could be both contribution on cellprob and contribution on gradient flows and w a weight)?

2) λ is set to 5, so are we giving more importance to gradient flows contribution or to ground-thruth wrt Cellpose output?

3) If this value is useful to give more importance to ground-thruth wrt Cellpose output, why did you set this weighted parameter to 5? Did you try several values? Did you discover what happens if this value is lower or higher?

## Notes

I did not find useful explanations in your papers and in issues opened/closed in your GitHub repository. In particular issues about "no detected objects" are related to small training dataset (it is not my case, since I'm dealing with at least 3000 samples per epoch) and to bugs.

**Comment by mrariden** [comment]
The answers to your questions are in the supplement of the paper which you can [find here](https://t.co/kBMXmPp3Yn?amp=1), on pages 8 and 9.

**Comment by ivanmag22** [comment]
Hi @mrariden , thank you for the answer, but I did not find a clarification there. It is repeated that it is used factor 5 for horizontal and vertical flows and in the supplements I did not find anything useful. Where can I find these answers? Thank you

**Comment by mrariden** [comment]
The supplement, on page 9 says, "To match the relative contributions of the L2 loss and cross-entropy loss, we multiplied the gradients by a factor of five". In other words, the constant is to make the two loss components relatively equal. 

The GT gradient values are normalized on [-1, 1], but the network produces any magnitude gradient values. Not including the 5 factor would unnecessarily constrain the predicted gradients and also make the gradient loss smaller in comparison to the BCE loss.

**Comment by ivanmag22** [comment]
Ok, maybe it is more clear. Can you clarify to me when flow normalization happens?

given *sample.tif* (our sample) and *sample_seg.npy* 
in **train.py**
```
train_seg():
  out = _process_train_test(train_labels=train_labels,)

_process_train_test():
  train_labels = [io.imread(train_labels_files[i]) for i in trange(nimg)]
```
then in **io.py**
```
imread():
  dat = np.load(filenmae, allow_pickle=True).item()
  masks = dat["masks"]
  return masks
```
then (again) in **train.py**
```
train_seg():
  train_labels = dybamics.labels_to_flows(train_labels, files=train_files, device=device)
```
so in **dynamics.py**
```
labels_to_flows():
  vecn = masks_to_flows(labels[n][0].astype(int), device=device, niter=niter)
  flow = np.concatenate((labels[n], labels[n] > 0.5, vecn), axis=0).astype(np.float32)

masks_to_flows():
  elif masks.ndim == 2:
  mu, mu_c = masks_to_flows_device(masks, device=device, niter=niter)
  return mu

masks_to_flows_cpu():
  T = np.zeros((ly) * (lx), np.float64)
  T = _extend_centers(T, y, x, ymed, xmed, np.int32(lx), np.int32(n_iter))
  dy = T[(y + 1) * lx + x] - T[(y - 1) * lx + x]
  dx = T[y * lx + x + 1] - T[y * lx + x - 1]
  mu[:, sr.start + y - 1, sc.start + x - 1] = np.stack((dy, dx))
  mu /= 1e-60 + (mu**2).sum(axis=0)**0.5
```

I want to understand if GT gradient values normalization in [-1,1] happens in this last method or somewhere else.
Thanks

**Comment by carsen-stringer** [comment]
flow normalization happens in the last line of masks_to_flows:   mu /= 1e-60 + (mu**2).sum(axis=0)**0.5
